// _rbx@1.1.1/event-stream-v2/src/client.ts
import { create, toBinary, type DescMessage } from "@bufbuild/protobuf";
import { gzipSync } from "fflate";

import { WebEventBase, WebEventBaseSchema } from "@rbx/event-stream-proto/event/web_event_base_pb.js";
import {
  SendEventRequestSchema,
  SendOptimizedBatchRequestSchema,
  SendOptimizedBatchRequest_SubBatchSchema
} from "@rbx/service-contracts-proto/roblox/experiencesignalsingest/experiencesignalsingest/v1/experience_signals_ingest_api_pb.js";

import type { Event, Message, Schema } from "./types.js";

interface EventStreamClientConfig {
  baseUrl?: string;
}

export class EventStreamClient {
  private baseUrl: string;

  constructor({
    baseUrl = 'https://apis.roblox.com',
  }: EventStreamClientConfig = {}) {
    this.baseUrl = baseUrl.replace(/\/$/, '') + '/experience-signals-ingest/public';
  }

  async sendEvent<T extends DescMessage>(schema: Schema<T>, msg: Message<T>): Promise<void> {
    try {
      const eventPayload = toBinary(schema, create(schema, msg), {writeUnknownFields: false});
      
      const req = create(SendEventRequestSchema, {
        event: {
          source: schema.typeName,
          payload: eventPayload,
        }
      });

      const response = await fetch(`${this.baseUrl}/v1/events/single`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/x-protobuf',
          'Content-Encoding': 'gzip',
        },
        body: gzipSync(toBinary(SendEventRequestSchema, req, {writeUnknownFields: false})) as Uint8Array<ArrayBuffer>,
        credentials: 'include',
        keepalive: true,
      });

      if (!response.ok) {
        throw new Error(`status ${response.status}`);
      }
    } catch (error) {
      console.error('EventStream failed to send event:', error);
    }
  }

  async sendBatch(events: Event[]): Promise<void> {
    try {
      const group = new Map<string, Event[]>();
      for (const event of events) {
        const source = event.schema.typeName;
        if (!group.has(source)) {
          group.set(source, []);
        }
        group.get(source)!.push(event);
      }

      const subBatches = Array.from(group.entries()).map(([source, entries]) => {
        const payloads = entries.map(({ schema, msg }) =>
          toBinary(schema, create(schema, msg), { writeUnknownFields: false })
        );

        return create(SendOptimizedBatchRequest_SubBatchSchema, {
          source,
          payloads,
        });
      });

      const req = create(SendOptimizedBatchRequestSchema, {
        subBatches
      });

      const response = await fetch(`${this.baseUrl}/v1/events/optimized`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/x-protobuf',
          'Content-Encoding': 'gzip',
        },
        body: gzipSync(toBinary(SendOptimizedBatchRequestSchema, req, {writeUnknownFields: false})) as Uint8Array<ArrayBuffer>,
        credentials: 'include',
        keepalive: true,
      });

      if (!response.ok) {
        throw new Error(`status ${response.status}`);
      }
    } catch (error) {
      console.error('EventStream failed to send batch events:', error);
    }
  }
}

export function webEventBase(params: {
  pageId?: string;
  referrerUrl?: string;
  sessionId?: string;
  guestId?: bigint;
  localTime?: Date;
} = {}): WebEventBase {
  return create(WebEventBaseSchema, {
    pageId: params.pageId ?? "",
    referrerUrl: params.referrerUrl ?? "",
    sessionId: params.sessionId ?? "",
    guestId: params.guestId ?? 0n,
    localTimestamp: (params.localTime ?? new Date()).toISOString()
  });
}
