// _rbx@4.2.0/badge-components/utils/encodeSvg.ts
// @ts-nocheck
const symbols = /[\r\n%#()<>?\[\\\]^`{|}]/g;

export const blankSvg =
  '<svg xmlns="http://www.w3.org/2000/svg" version="1.1" viewBox="0 0 500 500" width="500" height="500" id="starter_svg"></svg>';

export enum QuoteTypes {
  SINGLE = "single",
  DOUBLE = "double",
}

export const encodeSVG = (
  data: string,
  externalQuotesValue: QuoteTypes
): string => {
  let encodedData = data;
  // Use single quotes instead of double to avoid encoding.
  if (externalQuotesValue === QuoteTypes.DOUBLE) {
    encodedData = encodedData.replace(/"/g, "'");
  } else {
    encodedData = encodedData.replace(/'/g, '"');
  }

  encodedData = encodedData.replace(/>\s{1,}</g, "><");
  encodedData = encodedData.replace(/\s{2,}/g, " ");

  return encodedData.replace(symbols, encodeURIComponent);
};

export const getDataImageString = (escapedEncodedSvg: string) => {
  return `data:image/svg+xml;charset=utf-8,${escapedEncodedSvg}`;
};

export const svgToEncodedDataImageString = (svgMarkup: string) => {
  return getDataImageString(encodeSVG(svgMarkup, QuoteTypes.DOUBLE));
};
