// _rbx/identity-verification/dist/index.js
// src/QRDeepLinkDialog.tsx
import React, { useEffect, useState } from "react";
import { Dialog, DialogBody, DialogContent, DialogTitle } from "@rbx/foundation-ui";

// src/qrCode.ts
import { create as createQRCode } from "qrcode";
var generateQRCodeData = (deeplink) => {
  const qr = createQRCode(deeplink, { errorCorrectionLevel: "H" });
  const moduleCount = qr.modules.size;
  let pathData = "";
  for (let row = 0; row < moduleCount; row += 1) {
    for (let col = 0; col < moduleCount; col += 1) {
      if (qr.modules.get(row, col)) {
        pathData += `M${col} ${row}h1v1h-1z`;
      }
    }
  }
  return { pathData, moduleCount };
};

// src/QRDeepLinkDialog.tsx
import "./QRDeepLinkDialog-DANE25CR.css";
import { jsx, jsxs } from "react/jsx-runtime";
var QR_DISPLAY_SIZE_PX = 176;
var LOGO_OUTER_RATIO = 0.25;
var LOGO_OUTER_MIN_MODULES = 7;
var LOGO_FRAME_TARGET_PX = 5;
var computeLogoModules = (moduleCount) => {
  const target = Math.round(moduleCount * LOGO_OUTER_RATIO);
  const outer = Math.max(LOGO_OUTER_MIN_MODULES, target % 2 === 0 ? target - 1 : target);
  const targetFrame = Math.round(moduleCount * LOGO_FRAME_TARGET_PX / QR_DISPLAY_SIZE_PX);
  const maxFrame = Math.floor((outer - 5) / 2);
  const frame = Math.max(1, Math.min(maxFrame, targetFrame));
  const inner = outer - 2 * frame;
  return { outer, inner, glyph: inner - 2 };
};
var ROBLOX_LOGO_VIEWBOX_SIZE = 56;
var ROBLOX_LOGO_PATH_D = "M11.6763 0L0 44.1659L43.5771 56L55.2533 11.8341L11.6763 0ZM32.0849 35.827L19.9079 32.5185L23.1723 20.1769L35.3542 23.4855L32.0849 35.827Z";
var QRDeepLinkDialog = ({
  open,
  onOpenChange,
  deeplink,
  title,
  description,
  footer,
  closeAffordance,
  hideLogo = false
}) => {
  const [qrCode, setQrCode] = useState(null);
  useEffect(() => {
    try {
      setQrCode(generateQRCodeData(deeplink));
    } catch (err) {
      console.error("Failed to generate QRDeepLinkDialog QR code", err);
      setQrCode(null);
    }
  }, [deeplink]);
  if (qrCode === null) {
    return null;
  }
  const logo = computeLogoModules(qrCode.moduleCount);
  return /* @__PURE__ */ jsx(
    Dialog,
    {
      open,
      onOpenChange,
      isModal: true,
      size: "Medium",
      type: "Default",
      hasCloseAffordance: true,
      closeLabel: closeAffordance,
      children: /* @__PURE__ */ jsx(DialogContent, { children: /* @__PURE__ */ jsxs(DialogBody, { className: "flex flex-col", children: [
        /* @__PURE__ */ jsx(DialogTitle, { className: "text-heading-small padding-bottom-small", children: title }),
        /* @__PURE__ */ jsx("p", { className: "text-body-medium", children: description }),
        /* @__PURE__ */ jsx("div", { className: "flex justify-center", children: /* @__PURE__ */ jsx("div", { className: "iv-qr-frame padding-medium margin-y-large", children: /* @__PURE__ */ jsxs(
          "svg",
          {
            className: "iv-qr-svg",
            viewBox: `0 0 ${qrCode.moduleCount} ${qrCode.moduleCount}`,
            role: "img",
            "aria-label": title,
            children: [
              /* @__PURE__ */ jsx("path", { d: qrCode.pathData, fill: "#000000" }),
              !hideLogo && /* @__PURE__ */ jsxs(React.Fragment, { children: [
                /* @__PURE__ */ jsx(
                  "rect",
                  {
                    x: (qrCode.moduleCount - logo.outer) / 2,
                    y: (qrCode.moduleCount - logo.outer) / 2,
                    width: logo.outer,
                    height: logo.outer,
                    fill: "#FFFFFF"
                  }
                ),
                /* @__PURE__ */ jsx(
                  "rect",
                  {
                    className: "iv-qr-logo-backdrop",
                    x: (qrCode.moduleCount - logo.inner) / 2,
                    y: (qrCode.moduleCount - logo.inner) / 2,
                    width: logo.inner,
                    height: logo.inner,
                    fill: "#000000"
                  }
                ),
                /* @__PURE__ */ jsx(
                  "g",
                  {
                    transform: `translate(${(qrCode.moduleCount - logo.glyph) / 2} ${(qrCode.moduleCount - logo.glyph) / 2}) scale(${logo.glyph / ROBLOX_LOGO_VIEWBOX_SIZE})`,
                    children: /* @__PURE__ */ jsx("path", { d: ROBLOX_LOGO_PATH_D, fill: "#FFFFFF" })
                  }
                )
              ] })
            ]
          }
        ) }) }),
        /* @__PURE__ */ jsx("p", { className: "text-body-medium padding-bottom-small", children: footer })
      ] }) })
    }
  );
};
export {
  QRDeepLinkDialog,
  generateQRCodeData
};
