// _rbx@0.0.11/unified-logging/dist/index.js
import { useRef, useCallback, useMemo, useEffect } from 'react';
import debounce from 'lodash.debounce';
import 'intersection-observer';

// Polyfill for intersection-observer for older browsers
function observeVisibility(options, callback) {
    const { element, threshold } = options;
    try {
        const intersectionObserver = new IntersectionObserver(([entry]) => {
            callback(entry?.isIntersecting || false);
        }, { threshold });
        intersectionObserver.observe(element);
        return () => intersectionObserver.disconnect();
    }
    catch (_) {
        // Ignore cases where IntersectionObserver does not exist in older browsers, since we have added the polyfill
    }
    return () => undefined;
}
function observeChildrenVisibility(options, callback) {
    const { elements, threshold } = options;
    try {
        const intersectionObserver = new window.IntersectionObserver(callback, { threshold });
        elements.forEach(el => {
            intersectionObserver.observe(el);
        });
        return () => intersectionObserver.disconnect();
    }
    catch (_) {
        // Ignore cases where IntersectionObserver does not exist in older browsers, since we have added the polyfill
    }
    return () => undefined;
}
var elementVisibilityService = {
    observeVisibility,
    observeChildrenVisibility
};

const splitArray = (inputArray, chunkSize) => {
    if (inputArray.length === 0 || chunkSize === 0) {
        return [inputArray];
    }
    const result = [];
    for (let i = 0; i < inputArray.length; i += chunkSize) {
        result.push(inputArray.slice(i, i + chunkSize));
    }
    return result;
};

const impressionConstants = {
    // Include up to 25 items in each impressions event
    maxItemsInImpressionsEvent: 25,
    // When 50% or more of a tile is visible, send an impression event
    impressionsItemIntersectionThreshold: 0.5
};

const { maxItemsInImpressionsEvent, impressionsItemIntersectionThreshold } = impressionConstants;
const getSortedImpressedIndexes = (tileElements, entries, observer) => {
    const impressedIndexes = [];
    entries.forEach(entry => {
        if (entry?.isIntersecting) {
            const tileIndex = tileElements.findIndex(tileElement => tileElement === entry.target);
            if (tileIndex >= 0) {
                impressedIndexes.push(tileIndex);
                observer.unobserve(entry.target);
            }
        }
    });
    return impressedIndexes.sort((a, b) => a - b);
};
/**
 * Tracks the visibility of the children elements within the containerRef, and
 * calls onImpressionsDetected with the indexes of the children that have been impressed.
 *
 * Note: containerRef must be the direct parent of the children elements to track.
 */
const useCollectionItemsIntersectionTracker = (containerRef, childrenLength, onImpressionsDetected) => {
    const reportedImpressionsRef = useRef(new Set());
    const impressedRangeRef = useRef(new Set());
    const disconnectRef = useRef(null);
    const sendImpressions = useCallback(() => {
        const impressedRangeArray = Array.from(impressedRangeRef.current);
        const subArrays = splitArray(impressedRangeArray.filter(index => !reportedImpressionsRef.current.has(index)), maxItemsInImpressionsEvent).filter((indexes) => indexes.length > 0);
        subArrays.forEach((indexesToSend) => {
            onImpressionsDetected(indexesToSend);
            indexesToSend.forEach(index => reportedImpressionsRef.current.add(index));
        });
        impressedRangeRef.current.clear();
    }, [onImpressionsDetected]);
    const sendImpressionsDebounced = useMemo(() => {
        return debounce(() => sendImpressions(), 200);
    }, [sendImpressions]);
    useEffect(() => {
        const tileElements = Array.from(containerRef?.current?.children ?? []).filter((element) => element instanceof HTMLElement);
        const onObserve = (entries, observer) => {
            sendImpressionsDebounced.cancel();
            const sortedImpressedIndexes = getSortedImpressedIndexes(tileElements, entries, observer);
            sortedImpressedIndexes.forEach(index => impressedRangeRef.current.add(index));
            sendImpressionsDebounced();
        };
        // Disconnect the previous observer, if one exists
        if (disconnectRef.current) {
            disconnectRef.current();
            disconnectRef.current = null;
        }
        disconnectRef.current = observeChildrenVisibility({
            elements: tileElements,
            threshold: impressionsItemIntersectionThreshold
        }, onObserve);
        return () => {
            sendImpressionsDebounced.cancel();
            if (disconnectRef.current) {
                disconnectRef.current();
                disconnectRef.current = null;
            }
        };
    }, [containerRef, childrenLength, sendImpressionsDebounced, impressedRangeRef]);
};

var EventNames;
(function (EventNames) {
    EventNames["ItemImpressions"] = "itemImpressions";
    EventNames["ItemAction"] = "itemAction";
})(EventNames || (EventNames = {}));
var EventContext;
(function (EventContext) {
    EventContext["Games"] = "Games";
    EventContext["Home"] = "Home";
    EventContext["SearchLanding"] = "SearchLanding";
    EventContext["Spotlight"] = "Spotlight";
    EventContext["UserProfile"] = "UserProfile";
})(EventContext || (EventContext = {}));
var SessionInfo;
(function (SessionInfo) {
    SessionInfo["HomePageSessionInfo"] = "homePageSessionInfo";
    SessionInfo["DiscoverPageSessionInfo"] = "discoverPageSessionInfo";
    SessionInfo["GameSearchSessionInfo"] = "gameSearchSessionInfo";
    SessionInfo["SpotlightPageSessionInfo"] = "spotlightPageSessionInfo";
})(SessionInfo || (SessionInfo = {}));
var SharedEventMetadata;
(function (SharedEventMetadata) {
    SharedEventMetadata["ContentType"] = "contentType";
    SharedEventMetadata["Context"] = "context";
    SharedEventMetadata["CollectionId"] = "collectionId";
    SharedEventMetadata["CollectionPosition"] = "collectionPosition";
    SharedEventMetadata["CollectionComponentType"] = "collectionComponentType";
})(SharedEventMetadata || (SharedEventMetadata = {}));
var ContentType;
(function (ContentType) {
    ContentType["User"] = "User";
    ContentType["Game"] = "Game";
})(ContentType || (ContentType = {}));
var UserPresence;
(function (UserPresence) {
    UserPresence["Online"] = "online";
    UserPresence["InGame"] = "inGame";
    UserPresence["InStudio"] = "inStudio";
    UserPresence["Offline"] = "offline";
})(UserPresence || (UserPresence = {}));
var FriendStatus;
(function (FriendStatus) {
    FriendStatus["Friend"] = "friend";
    FriendStatus["NotFriend"] = "notFriend";
})(FriendStatus || (FriendStatus = {}));

var ItemImpressionsMetadata;
(function (ItemImpressionsMetadata) {
    ItemImpressionsMetadata["ItemIds"] = "itemIds";
    ItemImpressionsMetadata["ItemPositions"] = "itemPositions";
    ItemImpressionsMetadata["RowNumbers"] = "rowNumbers";
    ItemImpressionsMetadata["FeedRowNumbers"] = "feedRowNumbers";
    ItemImpressionsMetadata["PositionsInRow"] = "positionsInRow";
    ItemImpressionsMetadata["PositionsInTopic"] = "positionsInTopic";
    ItemImpressionsMetadata["TotalNumberOfItems"] = "totalNumberOfItems";
})(ItemImpressionsMetadata || (ItemImpressionsMetadata = {}));
var UserImpressionsMetadata;
(function (UserImpressionsMetadata) {
    UserImpressionsMetadata["Presences"] = "presences";
    UserImpressionsMetadata["PresenceUniverseIds"] = "presenceUniverseIds";
    UserImpressionsMetadata["FriendStatuses"] = "friendStatuses";
    UserImpressionsMetadata["SourceCarousel"] = "sourceCarousel";
})(UserImpressionsMetadata || (UserImpressionsMetadata = {}));

var ItemActionMetadata;
(function (ItemActionMetadata) {
    ItemActionMetadata["ItemId"] = "itemId";
    ItemActionMetadata["ItemPosition"] = "itemPosition";
    ItemActionMetadata["ItemComponentType"] = "itemComponentType";
    ItemActionMetadata["RowNumber"] = "rowNumber";
    ItemActionMetadata["FeedRowNumber"] = "feedRowNumber";
    ItemActionMetadata["PositionInRow"] = "positionInRow";
    ItemActionMetadata["PositionInTopic"] = "positionInTopic";
    ItemActionMetadata["TotalNumberOfItems"] = "totalNumberOfItems";
    ItemActionMetadata["ActionType"] = "actionType";
})(ItemActionMetadata || (ItemActionMetadata = {}));
var UserActionMetadata;
(function (UserActionMetadata) {
    UserActionMetadata["Presence"] = "presence";
    UserActionMetadata["PresenceUniverseId"] = "presenceUniverseId";
    UserActionMetadata["FriendStatus"] = "friendStatus";
    UserActionMetadata["SourceCarousel"] = "sourceCarousel";
})(UserActionMetadata || (UserActionMetadata = {}));

const parseEventParamValue = (value, shouldParseArraysWithoutQuotes // false for legacy implementation for existing events
) => {
    if (Array.isArray(value) && !shouldParseArraysWithoutQuotes) {
        return value.join(',');
    }
    if (typeof value === 'object' && value) {
        return JSON.stringify(value);
    }
    if (typeof value === 'number') {
        return value;
    }
    if (typeof value === 'string') {
        return value;
    }
    if (typeof value === 'boolean') {
        return value ? 1 : 0;
    }
    return undefined;
};
/**
 * Parses event params of an unknown type to a string or number for the event stream.
 *
 * Note: This function diverges from the parseEventParams() in web-frontend due to
 * the handling of arrays. The web-frontend version will add an extra set of quotation
 * marks around each string value due to the "JSON.stringify()", which results in
 * the downstream data stable having these extra quotation marks around each string.
 * However, existing events should use the legacy implementation below to avoid
 * breaking changes to pipelines that expect the extra quotation marks.
 */
const parseEventParams = (params) => {
    return Object.keys(params).reduce((acc, key) => {
        const parsedValue = parseEventParamValue(params[key], false);
        if (parsedValue !== undefined) {
            acc[key] = parsedValue;
        }
        return acc;
    }, {});
};

export { ContentType, EventContext, EventNames, FriendStatus, ItemActionMetadata, ItemImpressionsMetadata, SessionInfo, SharedEventMetadata, UserActionMetadata, UserImpressionsMetadata, UserPresence, elementVisibilityService, parseEventParams, useCollectionItemsIntersectionTracker };
