// _rbx@4.2.0/badge-components/containers/VerifiedBadgeTextContainer.tsx
import React, { FunctionComponent } from "react";
import classNames from "classnames";
import useVerifiedBadgeStyles from "../styles/VerifiedBadge.styles";
import { VerifiedBadgeIcon } from "../components/VerifiedBadgeIcon";
import { BadgeSizes } from "../constants/BadgeConstants";
import { noop, bindKeyDownA11yButton } from "../utils/misc";

// This property is not property supported with the MUI css properties type so we cannot include it in the VerifiedBadgeStyles code.
const defaultFontStyling = {
  fontWeight: 500,
};

export interface VerifiedBadgeTextContainerProps {
  size: BadgeSizes;
  titleText: string;
  showVerifiedBadge?: boolean;
  text?: string;
  textEl?: JSX.Element;
  badgeEl?: JSX.Element;
  overrideTextContainerClass?: string;
  overrideWrapperClass?: string;
  additionalTextContainerClass?: string;
  additionalWrapperClass?: string;
  onIconClick?: () => void;
}

export const verifiedBadgeTextWrapperDataAttrSelector =
  "data-rblx-badge-text-container";
export const verifiedBadgeTextElDataAttrSelector = "data-rblx-badge-text-el";
export const verifiedBadgeIconElDataAttrSelector = "data-rblx-badge-icon-el";
export const verifiedBadgeTextContainerReactRenderClass =
  "verified-badge-text-container";
const verifiedBadgeTextWrapperDataAttrs = {
  [verifiedBadgeTextWrapperDataAttrSelector]: "",
};
const verifiedBadgeTextElDataAttrs = {
  [verifiedBadgeTextElDataAttrSelector]: "",
};
const verifiedBadgeIconElDataAttrs = {
  [verifiedBadgeIconElDataAttrSelector]: "",
};

export const VerifiedBadgeTextContainer: FunctionComponent<
  VerifiedBadgeTextContainerProps
> = ({
  size,
  titleText,
  showVerifiedBadge = false,
  text,
  textEl = null,
  badgeEl = null,
  overrideTextContainerClass = "",
  overrideWrapperClass = "",
  additionalTextContainerClass = "",
  additionalWrapperClass = "",
  onIconClick = noop,
}) => {
  if (!size) {
    throw new Error("Must provide a size prop");
  }

  if (!titleText) {
    throw new Error("Must provide a titleText for accessibility");
  }

  const {
    wrapper = "",
    textContainer = "",
    badge = "",
    applyEllipsisTruncation = "",
  } = useVerifiedBadgeStyles(size).classes;

  const onKeyDownHandler = bindKeyDownA11yButton(onIconClick);
  const textContainerClasses: string =
    overrideTextContainerClass ||
    classNames(textContainer, additionalTextContainerClass);
  const wrapperClasses: string =
    overrideWrapperClass || classNames(wrapper, additionalWrapperClass);
  const textToRender = textEl || (
    <span className={applyEllipsisTruncation}>{text}</span>
  );
  const verifiedBadgeEl = showVerifiedBadge ? (
    <VerifiedBadgeIcon size={size} titleText={titleText} />
  ) : null;

  const finalBadgeElToRender = badgeEl || verifiedBadgeEl;

  const stylesToAddToTextContainer = overrideTextContainerClass
    ? {}
    : defaultFontStyling;

  return (
    <span {...verifiedBadgeTextWrapperDataAttrs} className={wrapperClasses}>
      <span
        {...verifiedBadgeTextElDataAttrs}
        style={stylesToAddToTextContainer}
        className={textContainerClasses}
      >
        {textToRender}
      </span>
      <span
        {...verifiedBadgeIconElDataAttrs}
        role="button"
        tabIndex={0}
        onClick={onIconClick}
        onKeyDown={onKeyDownHandler}
        className={badge}
      >
        {finalBadgeElToRender}
      </span>
    </span>
  );
};

export default VerifiedBadgeTextContainer;
