// _rbx@4.2.0/badge-components/styles/VerifiedBadge.styles.ts
import { makeStyles } from "tss-react/mui";
import { BadgeSizes } from "../constants/BadgeConstants";

const lineHeightMultiplier = 1.4;

export const SIZE_CONSTS: Record<BadgeSizes, Record<string, string>> = {
  [BadgeSizes.TITLE]: {
    marginLeft: "8px",
    fontSize: "28px",
    iconWidth: "28px",
    iconHeight: "28px",
  },
  [BadgeSizes.HEADER]: {
    marginLeft: "4px",
    fontSize: "24px",
    iconWidth: "16px",
    iconHeight: "16px",
  },
  [BadgeSizes.SUBHEADER]: {
    marginLeft: "4px",
    fontSize: "16px",
    iconWidth: "16px",
    iconHeight: "16px",
  },
  [BadgeSizes.CAPTIONHEADER]: {
    marginLeft: "4px",
    fontSize: "14px",
    iconWidth: "16px",
    iconHeight: "16px",
  },
  [BadgeSizes.FOOTER]: {
    marginLeft: "2px",
    fontSize: "12px",
    iconWidth: "12px",
    iconHeight: "12px",
  },
};

const useVerifiedBadgeStyles = (size: BadgeSizes) => {
  const verifiedBadgeStyles = {
    badge: {
      marginLeft: SIZE_CONSTS[size].marginLeft,
      position: "relative",
      display: "inline-block",
      flexGrow: 1,
    },

    iconImg: {
      height: SIZE_CONSTS[size].iconHeight,
      width: SIZE_CONSTS[size].iconWidth,
      marginBottom: "-3x",
    },

    textContainer: {
      maxWidth: `calc(100% - (${SIZE_CONSTS[size].iconWidth} + ${SIZE_CONSTS[size].marginLeft}))`,
      width: "min-content",
      fontStyle: "normal",
      textDecoration: "none",
      lineHeight: `${lineHeightMultiplier * 100}%`,
      fontSize: SIZE_CONSTS[size].fontSize,
      display: "inline-block",
      height: `calc(${SIZE_CONSTS[size].fontSize} * ${lineHeightMultiplier})`,
    },

    imgWrapper: {
      display: "flex",
      justifyContent: "flex-start",
      alignItems: "center",
      width: "auto",
    },

    wrapper: {
      position: "relative",
      display: "flex",
      justifyContent: "flex-start",
      alignItems: "center",
      width: "100%",
    },

    applyEllipsisTruncation: {
      width: "100%",
      display: "inline-block",
      whiteSpace: "nowrap",
      overflow: "hidden",
      textOverflow: "ellipsis",
    },
    // https://stackoverflow.com/questions/46710747/type-string-is-not-assignable-to-type-inherit-initial-unset-fixe
  } as const;
  return makeStyles()(verifiedBadgeStyles)();
};

export default useVerifiedBadgeStyles;
