// _rbx/translation-utils/dist/index.js
// src/translateHtml.tsx
import { Fragment } from "react";
import { jsx } from "react/jsx-runtime";
var FN_START_MARK = "__FN_nvfToKPAOuiV__";
var FN_END_MARK = "__FN_END_nvfToKPAOuiV__";
var FN_START = (key) => `${FN_START_MARK}${key}|`;
var FN_END = (key) => `${FN_END_MARK}${key}|`;
var FN_START_REGEXP = new RegExp(`${FN_START_MARK}(\\d+)\\|`);
var translateHtml = (translate, key, tags, args) => {
  const plainParams = { ...args };
  const reactSegments = {};
  tags?.forEach((tag, index) => {
    const reactKey = index.toString();
    const startMarker = FN_START(reactKey);
    const endMarker = FN_END(reactKey);
    plainParams[tag.opening] = startMarker;
    plainParams[tag.closing] = endMarker;
    reactSegments[reactKey] = {
      start: startMarker,
      end: endMarker,
      render: tag.render,
      used: false
    };
  });
  const intermediateTranslation = translate(key, plainParams);
  const generateTree = (translation) => {
    const parts = [];
    const reactSegmentStartMatch = FN_START_REGEXP.exec(translation);
    if (!reactSegmentStartMatch) {
      return [translation];
    }
    if (reactSegmentStartMatch.index > 0) {
      parts.push(translation.slice(0, reactSegmentStartMatch.index));
    }
    const reactInfo = reactSegmentStartMatch[1] && reactSegments[reactSegmentStartMatch[1]];
    if (!reactInfo) {
      console.warn("Unexpected malformed segment", key);
      return [];
    }
    reactInfo.used = true;
    const endIndex = translation.indexOf(reactInfo.end);
    if (endIndex === -1) {
      console.warn("Unexpected malformed segment", key);
      return [];
    }
    const subString = translation.slice(
      reactSegmentStartMatch.index + reactSegmentStartMatch[0].length,
      endIndex
    );
    const subStringResult = reactInfo.render(generateTree(subString));
    if (Array.isArray(subStringResult)) {
      parts.push(...subStringResult);
    } else {
      parts.push(subStringResult);
    }
    const endSegment = translation.slice(endIndex + reactInfo.end.length);
    if (endSegment.length > 0) {
      parts.push(...generateTree(endSegment));
    }
    return parts;
  };
  const result = generateTree(intermediateTranslation).filter((part) => part !== "");
  const hasUnusedSegments = Object.values(reactSegments).some((reactInfo) => !reactInfo.used);
  if (hasUnusedSegments) {
    console.warn("Unused segments found", key);
    return [];
  }
  return result.map((node, index) => /* @__PURE__ */ jsx(Fragment, { children: node }, index));
};
var translateHtml_default = translateHtml;

// src/translateHtmlLegacy.tsx
var translateHtmlLegacy = (translate, key, params) => {
  const tags = [];
  const args = {};
  params.forEach((param) => {
    if (param.length === 3) {
      const [opening, closing, render] = param;
      tags.push({ opening, closing, render });
    } else {
      const [paramKey, value] = param;
      args[paramKey] = value;
    }
  });
  return translateHtml_default(
    translate,
    key,
    tags.length > 0 ? tags : null,
    Object.keys(args).length > 0 ? args : void 0
  );
};
var translateHtmlLegacy_default = translateHtmlLegacy;
export {
  translateHtml_default as translateHtml,
  translateHtmlLegacy_default as translateHtmlLegacy
};
